File: ios/Runner/VisualObjectPlugin.swift
import Foundation
import Flutter
import ARKit
import UIKit

class VisualObjectPlugin: NSObject, FlutterPlugin {
    static var channel: FlutterMethodChannel?

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "live_captions_xr/visual_object_methods", binaryMessenger: registrar.messenger())
        VisualObjectPlugin.channel = channel
        let instance = VisualObjectPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // Example: Call this when you detect an ARAnchor or object
    static func sendVisualObjectDetected(anchor: ARAnchor, label: String, confidence: Double, boundingBox: [Double]) {
        let tf = anchor.transform
        let arr: [Double] = [
            Double(tf.columns.0.x), Double(tf.columns.0.y), Double(tf.columns.0.z), Double(tf.columns.0.w),
            Double(tf.columns.1.x), Double(tf.columns.1.y), Double(tf.columns.1.z), Double(tf.columns.1.w),
            Double(tf.columns.2.x), Double(tf.columns.2.y), Double(tf.columns.2.z), Double(tf.columns.2.w),
            Double(tf.columns.3.x), Double(tf.columns.3.y), Double(tf.columns.3.z), Double(tf.columns.3.w)
        ]
        let args: [String: Any] = [
            "label": label,
            "confidence": confidence,
            "boundingBox": boundingBox, // [left, top, right, bottom]
            "worldTransform": arr
        ]
        channel?.invokeMethod("onVisualObjectDetected", arguments: args)
    }

    // No-op: Dart->native calls not needed for this plugin
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(FlutterMethodNotImplemented)
    }
} 

File: android/app/src/main/kotlin/com/example/aurasense/VisualObjectPlugin.kt
package com.example.aurasense

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import com.google.ar.core.Anchor
import com.google.ar.core.Pose

class VisualObjectPlugin : MethodChannel.MethodCallHandler {
    companion object {
        private var channel: MethodChannel? = null

        fun registerWith(registrar: PluginRegistry.Registrar) {
            channel = MethodChannel(registrar.messenger(), "live_captions_xr/visual_object_methods")
            val instance = VisualObjectPlugin()
            channel?.setMethodCallHandler(instance)
        }

        // Example: Call this when you detect an Anchor or object
        fun sendVisualObjectDetected(anchor: Anchor, label: String, confidence: Double, boundingBox: List<Double>) {
            val pose: Pose = anchor.pose
            val matrix = FloatArray(16)
            pose.toMatrix(matrix, 0)
            val arr = matrix.map { it.toDouble() }
            val args = mapOf(
                "label" to label,
                "confidence" to confidence,
                "boundingBox" to boundingBox, // [left, top, right, bottom]
                "worldTransform" to arr
            )
            channel?.invokeMethod("onVisualObjectDetected", args)
        }
    }

    // No-op: Dart->native calls not needed for this plugin
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        result.notImplemented()
    }
} 